File: CS/ServicesMouseZoom/ServiceHandler.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraScheduler;
using DevExpress.XtraScheduler.Services;

namespace ServicesMouseZoom {
    public class MyMouseHandlerService : MouseHandlerServiceWrapper {
        IServiceProvider provider;
        bool myzoomEnabled;

        public MyMouseHandlerService(IServiceProvider provider, IMouseHandlerService service)
            : base(service) {
            this.provider = provider;
        }

        public override void OnMouseWheel(MouseEventArgs e) {
            DayView curView = ((SchedulerControl)provider).DayView;
            int nSlots = curView.TimeSlots.Count;

            if (this.myzoomEnabled) {
                if (e.Delta < 0) {
                    for (int i = 0; i < nSlots; i++) {
                        if (curView.TimeSlots[i].Value < curView.TimeScale) {
                            curView.TimeScale = curView.TimeSlots[i].Value;
                            break;
                        }
                    }

                }
                else {
                    for (int i = nSlots - 1; i >= 0; i--) {
                        if (curView.TimeSlots[i].Value > curView.TimeScale) {
                            curView.TimeScale = curView.TimeSlots[i].Value;
                            break;
                        }
                    }
                }
            }
            else {
                base.OnMouseWheel(e);
            }
        }

        public bool myZoomEnabled { get { return this.myzoomEnabled; } set { this.myzoomEnabled = value; } }


    }

    public class MyKeyboardHandlerService : KeyboardHandlerServiceWrapper {
        IServiceProvider provider;

        public MyKeyboardHandlerService(IServiceProvider provider, IKeyboardHandlerService service)
            : base(service) {
            this.provider = provider;

        }

        public override void OnKeyDown(KeyEventArgs e) {

            if (e.Control) {
                MyMouseHandlerService mouseHandler = (MyMouseHandlerService)provider.GetService(typeof(IMouseHandlerService));
                if (mouseHandler != null) {
                    mouseHandler.myZoomEnabled = !mouseHandler.myZoomEnabled;
                }
            }
            base.OnKeyDown(e);
        }
    }

}

File: VB/ServicesMouseZoom/ServiceHandler.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraScheduler
Imports DevExpress.Services
Imports DevExpress.Portable.Input

Namespace ServicesMouseZoom
	Public Class MyMouseHandlerService
		Inherits MouseHandlerServiceWrapper

		Private provider As IServiceProvider
'INSTANT VB NOTE: The field myzoomEnabled was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private myzoomEnabled_Conflict As Boolean

		Public Sub New(ByVal provider As IServiceProvider, ByVal service As IMouseHandlerService)
			MyBase.New(service)
			Me.provider = provider
		End Sub

		Public Overrides Sub OnMouseWheel(ByVal e As PortableMouseEventArgs)
			Dim curView As DayView = DirectCast(provider, SchedulerControl).DayView
			Dim nSlots As Integer = curView.TimeSlots.Count

			If Me.myzoomEnabled_Conflict Then
				If e.Delta < 0 Then
					For i As Integer = 0 To nSlots - 1
						If curView.TimeSlots(i).Value <= curView.TimeScale Then
							curView.TimeScale = curView.TimeSlots(i).Value
							Exit For
						End If
					Next i

				Else
					For i As Integer = nSlots - 1 To 0 Step -1
						If curView.TimeSlots(i).Value >= curView.TimeScale Then
							curView.TimeScale = curView.TimeSlots(i).Value
							Exit For
						End If
					Next i
				End If
			Else
				MyBase.OnMouseWheel(e)
			End If
		End Sub
		Public Property myZoomEnabled() As Boolean
			Get
				Return Me.myzoomEnabled_Conflict
			End Get
			Set(ByVal value As Boolean)
				Me.myzoomEnabled_Conflict = value
			End Set
		End Property
	End Class

	Public Class MyKeyboardHandlerService
		Inherits KeyboardHandlerServiceWrapper

		Private provider As IServiceProvider

		Public Sub New(ByVal provider As IServiceProvider, ByVal service As IKeyboardHandlerService)
			MyBase.New(service)
			Me.provider = provider
		End Sub
		Public Overrides Sub OnKeyDown(ByVal e As PortableKeyEventArgs)
			If e.Control Then
				Dim mouseHandler As MyMouseHandlerService = DirectCast(provider.GetService(GetType(IMouseHandlerService)), MyMouseHandlerService)
				If mouseHandler IsNot Nothing Then
					mouseHandler.myZoomEnabled = Not mouseHandler.myZoomEnabled
				End If
			End If
			MyBase.OnKeyDown(e)
		End Sub
	End Class

End Namespace
